internal/campaigns/volume_workspace.go

```diff
@@ -0,0 +1,199 @@
+		return errors.Wrap(err, "fetching repository archive")
+	}
+
+	// Create a Docker volume.
+	out, err := exec.CommandContext(ctx, "docker", "volume", "create").CombinedOutput()
+	if err != nil {
+		return errors.Wrap(err, "creating Docker volume")
+	}
+	w.volume = string(bytes.TrimSpace(out))
+
+	// Now mount that temporary file into a Docker container, and unzip it into
+	// the volume.
+	common, err := w.DockerRunOpts(ctx, "/work")
+	if err != nil {
+		return errors.Wrap(err, "generating run options")
+	}
+
+	opts := append([]string{
+		"run",
+		"--rm",
+		"--init",
+		"--workdir", "/work",
+		"--mount", "type=bind,source=" + hostZip + ",target=/tmp/zip,ro",
+	}, common...)
+	opts = append(opts, dockerWorkspaceImage, "unzip", "/tmp/zip")
+
+	out, err = exec.CommandContext(ctx, "docker", opts...).CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "unzip output:\n\n%s\n\n", string(out))
+	}
+
+	return nil
+}
+
+func (w *dockerVolumeWorkspace) Close(ctx context.Context) error {
+	// Cleanup here is easy: we just get rid of the Docker volume.
+	return exec.CommandContext(ctx, "docker", "volume", "rm", w.volume).Run()
+}
+
+func (w *dockerVolumeWorkspace) DockerRunOpts(ctx context.Context, target string) ([]string, error) {
+	return []string{
+		"--mount", "type=volume,source=" + w.volume + ",target=" + target,
+	}, nil
+}
+
+func (w *dockerVolumeWorkspace) WorkDir() *string { return nil }
+
+func (w *dockerVolumeWorkspace) Prepare(ctx context.Context) error {
+	script := `#!/bin/sh
+	
+set -e
+set -x
+
+git init
+# --force because we want previously "gitignored" files in the repository
+git add --force --all
+git commit --quiet --all -m src-action-exec
+`
+
+	if _, err := w.runScript(ctx, "/work", script); err != nil {
+		return errors.Wrap(err, "preparing workspace")
+	}
+	return nil
+}
+
+func (w *dockerVolumeWorkspace) Changes(ctx context.Context) (*StepChanges, error) {
+	script := `#!/bin/sh
+
+set -e
+# No set -x here, since we're going to parse the git status output.
+
+git add --all > /dev/null
+exec git status --porcelain
+`
+
+	out, err := w.runScript(ctx, "/work", script)
+	if err != nil {
+		return nil, errors.Wrap(err, "running git status")
+	}
+
+	changes, err := parseGitStatus(out)
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing git status output:\n\n%s", string(out))
+	}
+
+	return &changes, nil
+}
+
+func (w *dockerVolumeWorkspace) Diff(ctx context.Context) ([]byte, error) {
+	// As of Sourcegraph 3.14 we only support unified diff format.
+	// That means we need to strip away the `a/` and `/b` prefixes with `--no-prefix`.
+	// See: https://github.com/sourcegraph/sourcegraph/blob/82d5e7e1562fef6be5c0b17f18631040fd330835/enterprise/internal/campaigns/service.go#L324-L329
+	//
+	// Also, we need to add --binary so binary file changes are inlined in the patch.
+	script := `#!/bin/sh
+	
+exec git diff --cached --no-prefix --binary
+`
+
+	out, err := w.runScript(ctx, "/work", script)
```

Comment:
Not sure how you'll feel about this, I might be conflating implementation-detail and interface here, but since you're explicitly passing the script to `sh`, I think you don't need the shebang line, right?

```suggestion
	out, err := w.runScript(ctx, "/work", "exec git diff --cached --no-prefix --binary")
```

Reply:
You are correct: this was a stylistic choice rather than a technical one. I like having the shebang to reinforce that we're using Bourne shell (as opposed to bash) here, but I also accept that it makes the source code uglier. I'm not particularly tied to this, though — if you feel strongly about this, I'm happy to change it.

Reply:
Makes sense!